contracts/WildcardSteward_v2.sol

```diff
@@ -314,35 +322,37 @@ contract WildcardSteward_v2 is Initializable {
     }
 
     /* actions */
-    function _collectLoyalty(uint256 tokenId) internal {
-        // NOTE: this isn't currently implemented optimally. It would be possible to keep track of the total loyalty token generation rate for all the users tokens and use that.
-        // This should be implemented soon (while there are only a small number of tokens), or never.
-        address currentOwner = currentPatron[tokenId];
-        uint256 previousTokenCollection = timeLastCollected[tokenId];
-        uint256 patronageOwedByTokenPatron = patronageOwedPatron(currentOwner);
+    function _collectLoyaltyPatron(address tokenPatron) internal {
+        uint256 patronageOwedByTokenPatron = patronageOwedPatron(tokenPatron);
         uint256 timeSinceLastMint;
+        if (
+            patronageOwedByTokenPatron > 0 &&
+            patronageOwedByTokenPatron >= deposit[tokenPatron]
+        ) {
 
-        if (patronageOwedByTokenPatron >= deposit[currentOwner]) {
-            uint256 newTimeLastCollected = timeLastCollectedPatron[currentOwner]
-                .add(
+                uint256 previousCollectionTime
+             = timeLastCollectedPatron[tokenPatron];
+            // up to when was it actually paid for?
+            uint256 newTimeLastCollected = previousCollectionTime.add(
                 (
-                    (now.sub(timeLastCollectedPatron[currentOwner]))
-                        .mul(deposit[currentOwner])
+                    (now.sub(previousCollectionTime))
+                        .mul(deposit[tokenPatron])
                         .div(patronageOwedByTokenPatron)
                 )
             );
             timeSinceLastMint = (
-                newTimeLastCollected.sub(previousTokenCollection)
+                newTimeLastCollected.sub(previousCollectionTime)
             );
         } else {
-            timeSinceLastMint = now.sub(timeLastCollected[tokenId]);
+            timeSinceLastMint = now.sub(timeLastCollectedPatron[tokenPatron]);
         }
+
         mintManager.tokenMint(
-            currentOwner,
+            tokenPatron,
             timeSinceLastMint,
-            tokenGenerationRate[tokenId]
+            totalPatronTokenGenerationRate[tokenPatron]
```

Comment:
I cannot see anywhere, where this rate is updated to reflect if the `totalPatronTokenGenerationRate[tokenPatron]` decreases or increases, as the tokenPatron buys or has tokens bought from them?

Surely this needs to be addressed in the buy function. And everytime before this rate is changed, the _collectLoyaltyPatron needs to be called to mint all the previous outstanding tokens, before this rate is adjusted.

Reply:
Yeah, I hadn't pushed the latest changes... This PR is a bit of a mess because I rebased off your code. Lets get your code merged in; then this PR will definitely look cleaner.

But yes, it is handled now (if it wasn't before)